pkg/hive/job/job.go

```diff
@@ -0,0 +1,485 @@
+	shutdowner hive.Shutdowner
+
+	mu     lock.Mutex
+	groups []Group
+}
+
+func newRegistry(logger logrus.FieldLogger, shutdowner hive.Shutdowner) Registry {
+	return &registry{
+		logger:     logger,
+		shutdowner: shutdowner,
+	}
+}
+
+// NewGroup creates a new Group with the given `opts` options, which allows you to customize the behavior for the
+// group as a whole. For example by allowing you to add pprof labels to the group or by customizing the logger.
+func (c *registry) NewGroup(opts ...groupOpt) Group {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	var options options
+	options.logger = c.logger
+	options.shutdowner = c.shutdowner
+
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	g := &group{
+		options: options,
+
+		wg: &sync.WaitGroup{},
+	}
+
+	c.groups = append(c.groups, g)
+
+	return g
+}
+
+// Group aims to streamline the management of work within a cell. Group implements hive.HookInterface and takes care
+// of proper start and stop behavior as expected by hive. A group allows you to add multiple types of jobs which
+// different kinds of logic. No matter the job type, the function provided to is always called with a context which
+// is bound to the lifecycle of the cell.
+type Group interface {
+	Add(...Job)
+	hive.HookInterface
+}
+
+type Job interface {
+	start(ctx context.Context, wg *sync.WaitGroup, options options) error
+}
+
+type group struct {
+	options options
+
+	wg *sync.WaitGroup
+
+	mu         lock.Mutex
+	ctx        context.Context
+	cancel     context.CancelFunc
+	queuedJobs []Job
+}
+
+type options struct {
+	pprofLabels pprof.LabelSet
+	logger      logrus.FieldLogger
+	shutdowner  hive.Shutdowner
+}
+
+type groupOpt func(o *options)
+
+// WithLogger replaces the default logger with the given logger, useful if you want to add certain fields to the logs
+// created by the group/jobs.
+func WithLogger(logger logrus.FieldLogger) groupOpt {
+	return func(o *options) {
+		o.logger = logger
+	}
+}
+
+// WithPprofLabels adds pprof labels which will be added to the goroutines spawned for the jobs and thus included in
+// the pprof profiles.
+func WithPprofLabels(pprofLabels pprof.LabelSet) groupOpt {
+	return func(o *options) {
+		o.pprofLabels = pprofLabels
+	}
+}
+
+var _ hive.HookInterface = (*group)(nil)
+
+// Start implements the hive.HookInterface interface
+func (jg *group) Start(_ hive.HookContext) error {
+	jg.mu.Lock()
+	defer jg.mu.Unlock()
+
+	jg.ctx, jg.cancel = context.WithCancel(context.Background())
+
+	jg.wg.Add(len(jg.queuedJobs))
+	for _, job := range jg.queuedJobs {
+		pprof.Do(jg.ctx, jg.options.pprofLabels, func(ctx context.Context) {
+			go job.start(ctx, jg.wg, jg.options)
+		})
+	}
+
+	return nil
+}
+
+// Stop implements the hive.HookInterface interface
+func (jg *group) Stop(stopCtx hive.HookContext) error {
+	jg.mu.Lock()
+	defer jg.mu.Unlock()
+
+	done := make(chan struct{})
+	go func() {
+		jg.wg.Wait()
+		close(done)
+	}()
+
+	jg.cancel()
+
+	select {
+	case <-stopCtx.Done():
```

Comment:
I'm still not quite sure how we should handle the stop context. It feels wrong to silently give up and leave the resources around. It's a bit hard to say whether failing the stop when we couldn't clean up is less safe. Alternative is to ignore the stop context (as we're not doing a long running operation that we could "safely" abort) and just wait. If we overrun the timeout hive will forcefully stop. We probably should pick what semantics to recommend, but not quite sure yet since it somewhat depends on what other stop hooks might be doing and how critical it is that they run.

For now, let's at least log that we didn't manage to wait for jobs to stop and we're leaving them running.